crates/relayer/src/chain/ckb4ibc/monitor.rs

```diff
@@ -43,18 +45,33 @@ use crate::event::monitor::{Error, EventBatch, MonitorCmd, Next, Result, TxMonit
 use crate::event::IbcEventWithHeight;
 
 use super::cache_set::CacheSet;
-use super::utils::{generate_connection_id, get_prefix_search_key, get_script_hash};
+use super::utils::{
+    generate_connection_id, get_prefix_search_key, get_script_hash, tip_block_number,
+};
+
+#[derive(Eq, PartialOrd, Ord, PartialEq, Hash, Clone, Copy)]
+pub enum IbcProtocolType {
+    Connection,
+    Channel,
+    Packet,
+}
+
+pub type WriteAckMonitorSender = (Sender<Option<(IbcPacket, CellInput)>>, u64);
+pub type WriteAckMonitorCmd = Sender<WriteAckMonitorSender>;
 
 // TODO: add cell emitter here
 pub struct Ckb4IbcEventMonitor {
     rt: Arc<TokioRuntime>,
     rpc_client: Arc<RpcClient>,
     rx_cmd: Receiver<MonitorCmd>,
+    rx_write_ack: Receiver<WriteAckMonitorSender>,
     event_bus: EventBus<Arc<Result<EventBatch>>>,
     config: ChainConfig,
     cache_set: RwLock<CacheSet<H256>>,
     counterparty_client_type_rx: WatchReceiver<Option<ClientType>>,
     counterparty_client_type: ClientType,
+    fetch_cursors: Mutex<HashMap<IbcProtocolType, JsonBytes>>,
+    useless_write_ack_packets: Mutex<HashMap<u64, (IbcPacket, CellInput)>>,
```

Comment:
* Better add a comment what the key is (block number?).
* Use BTreeMap so it's easier to find packets < some block number?
* What if there are multiple packets in a block?

Reply:
It's also a bit strange that we are using both locking and message passing. Ideally we should share the map with locking and don't need message passing, or we don't need locking if we use message passing.

Reply:
> It's also a bit strange that we are using both locking and message passing. Ideally we should share the map with locking and don't need message passing, or we don't need locking if we use message passing.

use locking because these vars are running in a Tokio instance, and the instance is also running in monitor which runs in a single thread, and these vars also should pass message out of monitor thread, so I use Mutex to wrap

Reply:
Consider this a NIT. But tokio [`block_on`](https://docs.rs/tokio/latest/tokio/runtime/struct.Runtime.html#method.block_on) runs the future on the current thread, so I don't think these vars are sent/shared.

Reply:
I found the reason, I use Mutex to avoid adding `mut` to monitor self, because methods of `mut self` cannot be used in `Tokio::join!` macro